crates/flipperzero/src/storage.rs

```diff
@@ -0,0 +1,252 @@
+            },
+        )
+    }
+
+    /// Creates a new file. If file exist, truncate to zero size
+    pub fn create_always(self, set: bool) -> Self {
+        OpenOptions::from_parts(
+            self.access_mode,
+            if set {
+                self.open_mode | sys::FS_OpenMode_FSOM_CREATE_ALWAYS
+            } else {
+                self.open_mode & !sys::FS_OpenMode_FSOM_CREATE_ALWAYS
+            },
+        )
+    }
+
+    pub fn open(self, path: &CStr) -> Result<File, Error> {
+        // It's possible to produce a nonsensical `open_mode` using the above
+        // operations, so we have some logic here to drop any extraneous
+        // information. The possible open modes form a partial order (for
+        // example, `create_new` is more specialized than `truncate`) so we
+        // search for the first "on" bit in this sequence, and use that as the
+        // open mode.
+        let canonicalized_open_mode = if self.open_mode & sys::FS_OpenMode_FSOM_CREATE_NEW != 0 {
+            sys::FS_OpenMode_FSOM_CREATE_NEW
+        } else if self.open_mode & sys::FS_OpenMode_FSOM_CREATE_ALWAYS != 0 {
+            sys::FS_OpenMode_FSOM_CREATE_ALWAYS
+        } else if self.open_mode & sys::FS_OpenMode_FSOM_OPEN_APPEND != 0 {
+            sys::FS_OpenMode_FSOM_OPEN_APPEND
+        } else if self.open_mode & sys::FS_OpenMode_FSOM_OPEN_ALWAYS != 0 {
+            sys::FS_OpenMode_FSOM_OPEN_ALWAYS
+        } else {
+            sys::FS_OpenMode_FSOM_OPEN_EXISTING
+        };
+
+        let f = File::new();
+        if unsafe {
+            sys::storage_file_open(
+                f.0,
+                path.as_ptr() as *const i8,
+                self.access_mode,
+                canonicalized_open_mode,
+            )
+        } {
+            Ok(f)
+        } else {
+            // Per docs, "you need to close the file even if the open operation
+            // failed," but this is handled by `Drop`.
+            Err(Error::from_sys(unsafe { sys::storage_file_get_error(f.0) }))
+        }
+    }
+}
+
+/// File stream with buffered read operations.
+pub struct File(*mut sys::File);
+
+impl File {
+    pub fn new() -> Self {
+        unsafe {
+            File(sys::storage_file_alloc(
+                UnsafeRecord::open(RECORD_STORAGE).as_ptr(),
+            ))
+        }
+    }
+}
+
+impl Drop for File {
+    fn drop(&mut self) {
+        unsafe {
+            // `storage_file_close` calls `storage_file_sync`
+            // internally, so it's not necesssary to call it here.
+            sys::storage_file_close(self.0);
+        }
+    }
+}
+
+impl Read for File {
+    fn read(&mut self, buf: &mut [u8]) -> Result<usize, Error> {
+        let to_read = buf.len().try_into().map_err(|_| Error::InvalidParameter)?;
+        let bytes_read =
+            unsafe { sys::storage_file_read(self.0, buf.as_mut_ptr() as *mut c_void, to_read) };
+        if bytes_read == to_read || bytes_read < to_read && unsafe { sys::storage_file_eof(self.0) }
```

Comment:
Ah, good point. I think it makes more sense to check for an error unconditionally, rather than trying to infer what happened based on the number of bytes read.